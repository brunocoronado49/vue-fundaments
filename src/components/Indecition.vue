<template>
    <img v-if="img" :src="img" alt="bg">
    <div class="bg-dark"></div>

    <div class="indecition-container">
        <input
          type="text"
          v-model="question"
          v-on:keypress.enter="makeQuestion"
          placeholder="Ask me a question">
        <p>Remember finish it with "?" symbol</p>

        <div v-if="isValidQuestion">
            <h2>{{ question }}</h2>
            <h1>{{ answer }}</h1>
        </div>
    </div>
</template>

<script>
export default {
    name: 'Indecition',
    data() {
        return {
            question: null,
            answer: null,
            img: null,
            isValidQuestion: false
        }
    },
    watch: {
        question(value, oldValue) {
            this.isValidQuestion = false
            console.log({ value })

            if (!value.includes('?')) return
            this.isValidQuestion = true
            this.getAnswer()
        }
    },
    methods: {
        async getAnswer() {
            try {
                this.answer = 'Loading...'
                const {answer, image} = await fetch('https://yesno.wtf/api')
                    .then(res => res.json())
                
                this.answer = answer === 'yes' ? 'Si!' : 'No!'
                this.img = image
            } catch(error) {
                console.log('Indecision error:', error)
                this.answer = 'Cannot load api'
                this.img = null
            }
        }
    }
}
</script>

<style scoped>
    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecition-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
        margin-bottom: 10px;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }
</style>
