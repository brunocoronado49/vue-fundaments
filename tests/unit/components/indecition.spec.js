import { shallowMount } from '@vue/test-utils';
import Indecition from '@/components/Indecition'

describe('Indecision component test', () => {
    let wrapper;
    let clgSpy;

    global.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({ 
            answer: 'yes',
            forced: false,
            image: 'https://yesno.wtf/assets/yes/2.gif'
         })
    }))

    beforeEach(() => {
        wrapper = shallowMount(Indecition)
        clgSpy = jest.spyOn(console, 'log')
        jest.clearAllMocks()
    })

    test('should do match with snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Write in the input shouldn`t trigger nothing', async () => {
        const input = wrapper.find('input')
        await input.setValue('Unit test')
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')

        // Espia que el console haya sido llanado
        // expect(clgSpy).toHaveBeenCalledTimes(1)
        expect(clgSpy).toHaveBeenCalled()
        expect(getAnswerSpy).not.toHaveBeenCalled()
    })

    test('Trigger getAnswer when user write ? symbol', async () => {
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
        const input = wrapper.find('input')
        await input.setValue('Unit test?')
        expect(clgSpy).toHaveBeenCalled()
        expect(getAnswerSpy).toHaveBeenCalled()
    })

    test('getAnswer test', async () => {
        await wrapper.vm.getAnswer()
        const img = wrapper.find('img')
        expect(img.exists()).toBeTruthy()
        expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')
        expect(wrapper.vm.answer).toBe('Si!')

    })

    test('Test in getAnswer and api error', async () => {
        fetch.mockImplementationOnce(() => Promise.reject('API is down'))
        await wrapper.vm.getAnswer()
        const img = wrapper.find('img')
        expect(img.exists()).toBeFalsy()
        expect(wrapper.vm.answer).toBe('Cannot load api')
    })
})

