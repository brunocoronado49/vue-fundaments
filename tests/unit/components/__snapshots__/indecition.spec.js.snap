// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision component test should do match with snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecition-container"><input type="text" placeholder="Ask me a question">
  <p>Remember finish it with "?" symbol</p>
  <!--v-if-->
</div>
`;
